### PR TITLE
[PM-6523] generator service tuning

### PR DESCRIPTION
**Pull request by audrey-jensen**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [X] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Adjust credential generator interfaces to accommodate `SecretState` while avoiding race conditions arising from active user changes while an update is in progress.

**Note:** Most of this PR consists of mechanical code changes. It has a lot of files, but not much code changes in any of them. `DefaultGeneratorService` is the exception to this. Since it's where most user-specific operations occur, it basically was rewritten.

**Note:** There is commented code in this PR to fix a known issue once a upstream policy dependency becomes available (#7977). Swapping the commented code is out of scope of this PR.

**Note:** This PR does not need QA because this code is not yet consumed by user-facing code. (Soon though! The work it blocks is UI code!)

## Code changes

Platform changes:
- **libs/common/spec/index.ts:** Expose state fakes from the barrel file
- **libs/common/src/platform/state/index.ts:** Expose `CombinedState<T>`. For some reason, the typechecker fails to see that `SecretState.combinedState$` is structurally identical to `SingleUserState.combinedState$` without it.
- **libs/common/src/tools/generator/state/secret-state.ts:** `SecretState` needs to implement `SingleUserState` so that it can be returned by the generator strategies.

The generator itself needs `UserId` passed into its methods so that the stores it relies upon don't suddenly change to a different user when there's an account switch.
- **libs/common/src/tools/generator/abstractions/generator.service.abstraction.ts:** expose userId parameter
- **libs/common/src/tools/generator/default-generator.service.ts:** The default generator's cache moved to lookup time because it now depends on the user. The cache operates the same way, but is stashed in a `Map` to ensure that each user's stream remains independent.

The generator strategies need to return a fully-realized `SingleUserState` object so that encryption configuration can be decided by the strategy. (This is a prereq for the forwarder strategies.) They also need to account for the possibility that there is no active policy.
- **libs/common/src/tools/generator/abstractions/generator-strategy.abstraction.ts** 
- **libs/common/src/tools/generator/passphrase/passphrase-generator-strategy.ts** 
- **libs/common/src/tools/generator/password/password-generator-strategy.spec.ts** 
- **libs/common/src/tools/generator/username/catchall-generator-strategy.ts** 
- **libs/common/src/tools/generator/username/eff-username-generator-strategy.ts** 
- **libs/common/src/tools/generator/username/subaddress-generator-strategy.ts** 

Unit test updates to reflect behavior and interface changes:
- **libs/common/src/tools/generator/state/secret-state.spec.ts**
- **libs/common/src/tools/generator/default-generator.service.spec.ts:** This file also addresses the `FakeStateService` migration FIXME, since 90% of that code is now in `GeneratorStrategy.durableState`.
- **libs/common/src/tools/generator/passphrase/passphrase-generator-strategy.spec.ts** 
- **libs/common/src/tools/generator/passphrase/password-generator-strategy.spec.ts** 
- **libs/common/src/tools/generator/username/catchall-generator-strategy.spec.ts** 
- **libs/common/src/tools/generator/username/eff-username-generator-strategy.spec.ts** 
- **libs/common/src/tools/generator/username/subaddress-generator-strategy.spec.ts**

### CODE REVIEW

**Comment by audrey-jensen** [comment]
@MGibson1 - There's actually one more thing I missed, which I'm adding now. `SecretState` needs to implement `SingleUserState<T>`. It's a mild change, but the typechecker needs yet another type definition exported from the state store: `CombinedState<T>`. For some reason, it fails to see that the structures are identical without it. 😕

File: libs/common/src/tools/generator/default-generator.service.ts [hunk 1]
```diff
@@ -13,45 +13,57 @@ export class DefaultGeneratorService<Options, Policy> implements GeneratorServic
...
+    const evaluator$ = policies$.pipe(
+      map((policy) => this.strategy.evaluator(policy)),
+      share({
+        connector: () => new ReplaySubject(1),
+        resetOnRefCountZero: () => timer(this.strategy.cache_ms),
+      }),
+    );
+
+    return evaluator$;
   }
```
Comment (ttalty) [inline libs/common/src/tools/generator/default-generator.service.ts:62]: :question: Does `createEvaluator(userId: UserId)` Need the userId parameter for later work? It's not used in this private function.

**Review by ttalty — APPROVED** [review]
Approved.

File: libs/common/src/tools/generator/default-generator.service.ts (hunk 1 above)
Comment (audrey-jensen) [reply libs/common/src/tools/generator/default-generator.service.ts:62]: Correct. The userId is a required parameter for the `getAll$` function.

**Review by MGibson1 — APPROVED** [review]

**Pull request merged**